src/tcp_sock.rs

```diff
@@ -2,39 +2,47 @@ use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
 use maidsafe_utilities::serialisation::{deserialise_from, serialise_into};
 use mio::tcp::TcpStream;
 use mio::{Evented, Poll, PollOpt, Ready, Token};
+use out_queue::OutQueue;
 use serde::de::DeserializeOwned;
 use serde::ser::Serialize;
-use std::collections::{BTreeMap, VecDeque};
 use std::fmt::{self, Debug, Formatter};
 use std::io::{self, Cursor, ErrorKind, Read, Write};
 use std::mem;
 use std::net::{Shutdown, SocketAddr};
 use std::time::Duration;
-use std::time::Instant;
-use {Priority, SocketError, MAX_MSG_AGE_SECS, MAX_PAYLOAD_SIZE, MSG_DROP_PRIORITY};
+use {Priority, SocketConfig, SocketError};
 
+/// TCP socket which by default is uninitialized.
 pub struct TcpSock {
     inner: Option<Inner>,
 }
 
 impl TcpSock {
-    pub fn connect(addr: &SocketAddr) -> ::Res<Self> {
+    /// Starts TCP connection and uses given socket configuration. Returns immediately, so make
+    /// sure to wait until the socket becomes writable.
+    pub fn connect_with_conf(addr: &SocketAddr, conf: SocketConfig) -> ::Res<Self> {
         let stream = TcpStream::connect(addr)?;
-        Ok(Self::wrap(stream))
+        Ok(Self::wrap_with_conf(stream, conf))
     }
 
-    pub fn wrap(stream: TcpStream) -> Self {
+    /// Starts TCP connection. Returns immediately, so make sure to wait until the socket becomes
+    /// writable.
+    pub fn connect(addr: &SocketAddr) -> ::Res<Self> {
+        Self::connect_with_conf(addr, SocketConfig::default())
+    }
+
+    /// Wraps `TcpStream` and uses given socket configuration.
+    pub fn wrap_with_conf(stream: TcpStream, conf: SocketConfig) -> Self {
         TcpSock {
-            inner: Some(Inner {
-                stream,
-                read_buffer: Vec::new(),
-                read_len: 0,
-                write_queue: BTreeMap::new(),
-                current_write: None,
-            }),
+            inner: Some(Inner::new(stream, conf)),
```

Comment:
Having multiple constructors, one with default values, etc. for `Inner` just adds more boilerplate code. Since `Inner` is a private type and it always consumes socket config I would refrain from such changes.

Reply:
Ya but from reading the code p.o.v I want the defaults to be enforced by the types closest to constructing them which here is the `Inner` type. It's the same reason i'm saying that moving the construction to `Inner` as opposed to the outer doing it, was a good move.